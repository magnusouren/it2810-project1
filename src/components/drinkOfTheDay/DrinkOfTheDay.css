.drink-card {
  display: flex;
  flex-direction: column;
  border: 1px solid #ccc;
  border-radius: 10px;
  padding: 10px;
  margin: 40px;
  background-color: #f08080;
  color: white;
  transition: transform 0.2s ease;

  &:hover {
    transform: scale(1.03);
  }

  .link-unstyled {
    text-decoration: none;
    color: inherit;
  }

  .left {
    display: flex;
    flex-direction: column;
    align-items: center;
    margin-bottom: 10px;

    img {
      max-width: 100%;
      max-height: 150px;
      object-fit: cover;
    }
  }

  .right {
    display: flex;
    flex-direction: column;
    align-items: flex-start;

    ul {
      list-style-type: none;
      padding: 0;
    }
  }
}

@media only screen and (min-width: 968px) {
  .drink-card {
    flex-direction: row; /* Horizontal layout for larger screens */
    align-items: center; /* Center vertically on larger screens */
    .left {
      margin-right: 20px;
    }
  }
}
