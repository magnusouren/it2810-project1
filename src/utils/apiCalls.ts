import axios from 'axios';

export const fetchCocktailById = async (id: string) => {
  const response = await axios.get(`https://www.thecocktaildb.com/api/json/v1/1/lookup.php?i=${id}`);
  return response.data;
};
