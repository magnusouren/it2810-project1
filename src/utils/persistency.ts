import { DrinkOfTheDay } from '../types';

const addFavorite = (id: string) => {
  const favorites = getFavorites();
  favorites.push(id);
  localStorage.setItem('drinkFavorites', JSON.stringify(favorites));
};

export const getFavorites = () => {
  const favorites = localStorage.getItem('drinkFavorites');
  if (favorites) {
    return JSON.parse(favorites);
  }
  return [];
};

const removeFavorite = (id: string) => {
  const favorites = getFavorites();
  const newFavorites = favorites.filter((favorite: string) => favorite !== id);
  localStorage.setItem('drinkFavorites', JSON.stringify(newFavorites));
};

export const toggleFavorite = (id: string) => {
  const favorites = getFavorites();
  if (favorites.includes(id)) {
    removeFavorite(id);
  } else {
    addFavorite(id);
  }
};

export const isFavorite = (id: string) => {
  const favorites = getFavorites();
  return favorites.includes(id);
};

export const getDrinkOfTheDay = (date: string) => {
  const drinkOfTheDay = localStorage.getItem(date);
  if (drinkOfTheDay) {
    return JSON.parse(drinkOfTheDay);
  }
  return null;
};

export const setDrinkOfTheDay = (date: string, drink: DrinkOfTheDay) => {
  localStorage.setItem(date, JSON.stringify(drink));
};
