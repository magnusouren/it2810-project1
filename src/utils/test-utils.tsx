import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { render } from '@testing-library/react';
import { BrowserRouter, MemoryRouter, Route, Routes } from 'react-router-dom';

import { DrinkCard } from '../components/drink/DrinkCard';

// Function to render a component wrapped inside a browserrouter for testing purposes
export const renderWithRouter = (component: JSX.Element) => render(<BrowserRouter>{component}</BrowserRouter>);

const queryClient = new QueryClient();

export const renderWithRouterAndQueryClient = (component: JSX.Element) =>
  render(
    <QueryClientProvider client={queryClient}>
      <BrowserRouter>{component}</BrowserRouter>
    </QueryClientProvider>,
  );

export const renderWithRouterQueryClientAndDrinkId = (id: string) =>
  render(
    <QueryClientProvider client={new QueryClient()}>
      <MemoryRouter initialEntries={[`/drink/${id}`]}>
        <Routes>
          <Route path='/drink'>
            <Route path=':id' element={<DrinkCard />} />
          </Route>
        </Routes>
      </MemoryRouter>
    </QueryClientProvider>,
  );
