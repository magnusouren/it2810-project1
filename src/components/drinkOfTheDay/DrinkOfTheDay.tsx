import React, { useState, useEffect } from 'react';
import axios from 'axios';
import { Link } from 'react-router-dom';
import './DrinkOfTheDay.css';
interface Drink {
  strDrink: string;
  strDrinkThumb: string;
  strCategory: string;
  strInstructions: string;
  ingredients: string[];
}

export const DrinkOfTheDay: React.FC = () => {
  const [drink, setDrink] = useState<Drink | null>(null);

  useEffect(() => {
    // Fetch a random drink from the API
    axios
      .get('https://www.thecocktaildb.com/api/json/v1/1/random.php')
      .then((response) => {
        const randomDrink = response.data.drinks[0];
        const ingredients: string[] = [];

        // Extract ingredients
        for (let i = 1; i <= 15; i++) {
          const ingredient = randomDrink[`strIngredient${i}`];
          if (ingredient) {
            ingredients.push(`${ingredient}`);
          } else {
            break;
          }
        }

        randomDrink.ingredients = ingredients;
        setDrink(randomDrink);
      })
      .catch((error) => {
        console.error('Error fetching drink:', error);
      });
  }, []);

  return (
    <div className='drink-card'>
      {drink && (
        <Link to={`/categories/${drink.strCategory}`} className='link-unstyled'>
          <div className='left'>
            <h1>Drink of the day!</h1>
            <h2>{drink.strDrink}</h2>
            <img src={drink.strDrinkThumb + '/preview'} alt={drink.strDrink} />
          </div>
          <div className='right'>
            <p>Category: {drink.strCategory}</p>
            <p>Ingredients:</p>
            <ul>
              {drink.ingredients.map((ingredient, index) => (
                <li key={index}>{ingredient}</li>
              ))}
            </ul>
            <p>Instructions: {drink.strInstructions}</p>
          </div>
        </Link>
      )}
    </div>
  );
};
