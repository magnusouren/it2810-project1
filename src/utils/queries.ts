import axios, { isCancel } from 'axios';

import { Drink, DrinkOfTheDay, Ingredient } from '../types';
import { setDrinkOfTheDay } from './persistency';

export const fetchDrinkById = async (id?: string) => {
  if (!id) {
    return null;
  }
  const response = await axios
    .get(`https://www.thecocktaildb.com/api/json/v1/1/lookup.php?i=${id}`)
    .then((response) => {
      // Extract drink data
      const drinkData = response.data.drinks[0];
      const ingredients: Array<Ingredient> = [];
      const drink: Drink = {
        idDrink: drinkData.idDrink,
        strDrink: drinkData.strDrink,
        strGlass: drinkData.strGlass,
        ingredients: [],
        strInstructions: drinkData.strInstructions,
        strDrinkThumb: drinkData.strDrinkThumb,
        strCategory: drinkData.strCategory,
        strAlcoholic: drinkData.strAlcoholic,
      };

      // Extract ingredients
      for (let i = 1; i <= 15; i++) {
        const ingredient: string = drinkData[`strIngredient${i}`];
        const measure: string = drinkData[`strMeasure${i}`];
        if (ingredient) {
          ingredients.push({ ingredient, measure });
        } else {
          break;
        }
      }

      drink.ingredients = ingredients;
      return drink;
    })
    .catch((error) => {
      if (isCancel(error)) {
        console.log('Request cancelled', error.message);
        return null;
      }
      if (error instanceof TypeError) {
        return null;
      } else {
        console.error('Error fetching drink:', error);
        return null;
      }
    });
  return response;
};

export const fetchDrinkOfTheDay = async (currentDate: string) => {
  return axios
    .get('https://www.thecocktaildb.com/api/json/v1/1/random.php')
    .then((response) => {
      const randomDrink = response.data.drinks[0];

      const drink: DrinkOfTheDay = {
        drinkId: randomDrink.idDrink,
        strDrink: randomDrink.strDrink,
        strDrinkThumb: randomDrink.strDrinkThumb,
        strCategory: randomDrink.strCategory,
        strGlass: randomDrink.strGlass,
        strAlcoholic: randomDrink.strAlcoholic,
      };

      // Store the new random drink ID in local storage with the current date
      setDrinkOfTheDay(currentDate, drink);

      return drink;
    })
    .catch((error) => {
      console.error('Error fetching drink:', error);
      return null;
    });
};
