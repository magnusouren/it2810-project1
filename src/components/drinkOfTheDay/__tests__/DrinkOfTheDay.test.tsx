import { fireEvent, screen, waitFor } from '@testing-library/react';

import { renderWithRouterAndQueryClient } from '../../../utils/test-utils';
import { DrinkOfTheDay } from '../DrinkOfTheDay';

describe('DrinkOfTheDay', () => {
  it('Should match snapshot', async () => {
    const { asFragment } = renderWithRouterAndQueryClient(<DrinkOfTheDay />);

    await waitFor(() => {
      expect(screen.getByText('Drink of the day!')).toBeDefined();
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('Should render correct heading element', async () => {
    renderWithRouterAndQueryClient(<DrinkOfTheDay />);
    await waitFor(() => {
      expect(screen.getByText('Drink of the day!')).toBeDefined();
    });
    expect(screen.getByRole('heading', { name: 'Drink of the day!' })).toBeDefined();
  });

  it('Should render correct heading element for drink', async () => {
    renderWithRouterAndQueryClient(<DrinkOfTheDay />);
    await waitFor(() => {
      expect(screen.getByText('Random Drink')).toBeDefined();
    });
    expect(screen.getByRole('heading', { name: 'Random Drink' })).toBeDefined();
  });

  it('Should render correct details for drink', async () => {
    renderWithRouterAndQueryClient(<DrinkOfTheDay />);
    await waitFor(() => {
      expect(screen.getByText('Random Drink')).toBeDefined();
    });
    expect(screen.getByText('Category')).toBeDefined();
    expect(screen.getByText('Highball')).toBeDefined();
    expect(screen.getByText('Alcoholic')).toBeDefined();
  });

  it('Should render correct image for drink', async () => {
    const { getByAltText } = renderWithRouterAndQueryClient(<DrinkOfTheDay />);
    await waitFor(() => {
      expect(screen.getByText('Random Drink')).toBeDefined();
    });
    expect(getByAltText('Random Drink')).toBeDefined();
  });

  it('should navigate to correct drink on click', async () => {
    const { getByText } = renderWithRouterAndQueryClient(<DrinkOfTheDay />);
    fireEvent.click(getByText('Random Drink'));
    expect(window.location.pathname).toBe('/drink/9118');
  });
});
