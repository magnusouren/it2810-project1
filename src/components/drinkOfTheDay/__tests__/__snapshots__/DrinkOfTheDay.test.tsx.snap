// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DrinkOfTheDay > Should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="drink-of-the-day-card"
  >
    <div
      class="drink-of-the-day-card-container"
    >
      <a
        class="link-wrapper"
        href="/drink/9118"
      >
        <h2>
          Drink of the day!
        </h2>
        <div
          class="left"
        >
          <h3>
            Random Drink
          </h3>
          <div
            class="info-container"
          >
            <p
              id="drink-of-the-day-category"
            >
              Category
            </p>
            <p
              id="drink-of-the-day-glass"
            >
              Highball
            </p>
            <p
              id="drink-of-the-day-alcoholic"
            >
              Alcoholic
            </p>
          </div>
        </div>
        <div
          class="right"
        >
          <img
            alt="Random Drink"
            src="https://localhost:3000/drink/11118.jpg/preview"
          />
        </div>
      </a>
    </div>
  </div>
</DocumentFragment>
`;
